File: cklad.cpp
#include <iostream> 
#include <map> 
#include <vector>
#include <string>
#include <iomanip>
#include <locale>

using namespace std;

// Структура о товаре в ячейке - название и кол-во
struct Cell {
    string product;
    int quantity;
};

// склад map ключ - адрес ячейки, значение - вектор товаров {{A111 {{"name", count}, {}, {}}}, {}, {}}
map<string, vector<Cell>> store;

// количества товаров в ячейкe
int Quantity(const string& address) {
    int count = 0;
    const vector<Cell>& cellVector = store[address];  // Получение вектора товаров по адресу
    for (size_t i = 0; i < cellVector.size(); ++i) {  // Перебор всех товаров в ячейке
        count += cellVector[i].quantity;  // + кол-во текущего товара
    }
    return count;
}

// инициализации склада
void Creation() {
    vector<string> area = {"A", "B"};
    for (const string& area : area) {
        for (int rack = 1; rack <= 6; ++rack) {  // стеллажи - 6
            for (int section = 1; section <= 4; ++section) {  // секции - 4
                for (int shelf = 1; shelf <= 4; ++shelf) {  // полоки  - 4
                    string address = area + to_string(rack) + to_string(section) + to_string(shelf);  // адреса ячейки
                    store[address] = {};  // Инициализация пустой ячейки
                }
            }
        }
    }
}

// добавления товара
void add(const string& product, int quantity, const string& address) {
    if (store.count(address) == 0) {  // Проверка существования адреса
        cout << "Такого адреса не существует" << endl;
        return;
    }

    int can = 10 - Quantity(address);  // оставшееся место
    if (quantity > can) {
        cout << "По данному адресу вы можете положить не более " << can << " единиц товара." << endl;
        return;
    }

    vector<Cell>& cellVector = store[address];
    // есть ли уже такой товар в ячейке
    for (size_t i = 0; i < cellVector.size(); ++i) {
        Cell& cell = cellVector[i];
        if (cell.product == product) {
            cell.quantity += quantity;
            return;
        }
    }

    // Добавление нового товара в ячейку
    cellVector.push_back({product, quantity});
}

// удаления товара из ячейки
void remove(const string& product, int quantity, const string& address) {
    if (store.count(address) == 0) {
        cout << "Такого адреса не существует" << endl;
        return;
    }

    vector<Cell>& cellVector = store[address];
    for (size_t i = 0; i < cellVector.size(); ++i) {
        Cell& cell = cellVector[i]; 
        if (cell.product == product) {
            if (cell.quantity < quantity) {
                cout << "Вы удаляете больше единиц товара, чем есть в ячейке" << endl;
                return;
            }
            cell.quantity -= quantity;
            if (cell.quantity == 0) {
                cellVector.erase(cellVector.begin() + i);  // Удаление товара из ячейки, если его количество стало 0
                return;
            }
            return;
        }
    }

    cout << "Такого товара нет, чтобы его удалять" << endl;
}

// Функция для вывода информации о состоянии склада
void info() {
    int totalCapacity = 1920;  // Общая вместимость склада
    int totalItems = 0;         // Общее количество товаров на складе
    int zoneACapacity = 960;   // Вместимость зоны A
    int zoneBCapacity = 960;   
    int zoneAItems = 0;        // Количество товаров в зоне A
    int zoneBItems = 0;        

    cout << "Состояние склада:" << endl;
    for (map<string, vector<Cell>>::const_iterator it = store.begin(); it != store.end(); ++it) {  // Перебор всех ячеек
        const string& address = it->first;  //  адреса
        const vector<Cell>& cellVector = it->second;  // вектора товаров
        if (!cellVector.empty()) { 
            cout << address << ": ";
            for (size_t i = 0; i < cellVector.size(); ++i) {  // Перебор всех товаров в ячейке
                const Cell& cell = cellVector[i]; 
                if (cell.quantity > 0) {
                    cout << cell.product << " (" << cell.quantity << " шт.) ";
                    totalItems += cell.quantity;
                    if (address[0] == 'A') {
                        zoneAItems += cell.quantity;  // Увеличение количества товаров в зоне A
                    } else if (address[0] == 'B') {
                        zoneBItems += cell.quantity;
                    }
                }
            }
            cout << endl;
        }
    }

    cout << "Пустые ячейки:" << endl;
    for (map<string, vector<Cell>>::const_iterator it = store.begin(); it != store.end(); ++it) {
        if (it->second.empty() || Quantity(it->first) == 0) { 
            cout << it->first << " "; 
        }
    }

    cout << endl;

    //насколько процентов загружен склад, на сколько процентов загружена каждая зона склада
    cout << "Склад загружен на: " << fixed << setprecision(2)
         << (static_cast<double>(totalItems) / totalCapacity) * 100 << "%" << endl;
    cout << "Зона A загружена на: " << (static_cast<double>(zoneAItems) / zoneACapacity) * 100 << "%" << endl;
    cout << "Зона B загружена на: " << (static_cast<double>(zoneBItems) / zoneBCapacity) * 100 << "%" << endl;
}

int main() {
    Creation();  // Инициализация склада
    string command, product, address;
    int quantity;

    while (true) {
        cin >> command; 
        if (command == "ADD") { 
            cin >> product >> quantity >> address;
            add(product, quantity, address);
        } else if (command == "REMOVE") { 
            cin >> product >> quantity >> address;
            remove(product, quantity, address); 
        } else if (command == "INFO") {
            info();
        } else if (command == "EXIT") {
            break;
        } else {
            cout << "Такой команды не существует" << endl;
        }
    }

    return 0;
}

File: go/cklad.go
package main

import (
	"fmt"
	"strings"
)

// Структура для хранения информации о товаре в ячейке
type Cell struct {
	product  string
	quantity int
}

// Глобальная переменная для хранения состояния склада
var store = make(map[string][]Cell)

// Функция для получения количества товаров в ячейке
func Quantity(address string) int {
	count := 0
	cellVector := store[address]
	for _, cell := range cellVector {
		count += cell.quantity
	}
	return count
}

// Функция для инициализации склада
func Creation() {
	areas := []string{"A", "B"}
	for _, area := range areas {
		for rack := 1; rack <= 6; rack++ {
			for section := 1; section <= 4; section++ {
				for shelf := 1; shelf <= 4; shelf++ {
					address := fmt.Sprintf("%s%d%d%d", area, rack, section, shelf)
					store[address] = []Cell{}
				}
			}
		}
	}
}

// Функция для добавления товара в ячейку
func add(product string, quantity int, address string) {
	if _, exists := store[address]; !exists {
		fmt.Println("Такого адреса не существует")
		return
	}

	can := 10 - Quantity(address)
	if quantity > can {
		fmt.Printf("По данному адресу вы можете положить не более %d единиц товара.\n", can)
		return
	}

	cellVector := store[address]
	for i := range cellVector {
		if cellVector[i].product == product {
			cellVector[i].quantity += quantity
			return
		}
	}

	store[address] = append(cellVector, Cell{product, quantity})
}

// Функция для удаления товара из ячейки
func remove(product string, quantity int, address string) {
	if _, exists := store[address]; !exists {
		fmt.Println("Такого адреса не существует")
		return
	}

	cellVector := store[address]
	for i := range cellVector {
		if cellVector[i].product == product {
			if cellVector[i].quantity < quantity {
				fmt.Println("Вы удаляете больше единиц товара, чем есть в ячейке")
				return
			}
			cellVector[i].quantity -= quantity
			if cellVector[i].quantity == 0 {
				store[address] = append(cellVector[:i], cellVector[i+1:]...)
			}
			return
		}
	}

	fmt.Println("Такого товара нет, чтобы его удалять")
}

// Функция для вывода информации о состоянии склада
func info() {
	totalCapacity := 1920
	totalItems := 0
	zoneACapacity := 960
	zoneBCapacity := 960
	zoneAItems := 0
	zoneBItems := 0

	fmt.Println("Состояние склада:")
	for address, cellVector := range store {
		if len(cellVector) > 0 {
			fmt.Printf("%s: ", address)
			for _, cell := range cellVector {
				if cell.quantity > 0 {
					fmt.Printf("%s (%d шт.) ", cell.product, cell.quantity)
					totalItems += cell.quantity
					if strings.HasPrefix(address, "A") {
						zoneAItems += cell.quantity
					} else if strings.HasPrefix(address, "B") {
						zoneBItems += cell.quantity
					}
				}
			}
			fmt.Println()
		}
	}

	fmt.Println("Пустые ячейки:")
	for address := range store {
		if len(store[address]) == 0 || Quantity(address) == 0 {
			fmt.Printf("%s ", address)
		}
	}
	fmt.Println()

	fmt.Printf("Склад загружен на: %.2f%%\n", float64(totalItems)/float64(totalCapacity)*100)
	fmt.Printf("Зона A загружена на: %.2f%%\n", float64(zoneAItems)/float64(zoneACapacity)*100)
	fmt.Printf("Зона B загружена на: %.2f%%\n", float64(zoneBItems)/float64(zoneBCapacity)*100)
}

func main() {
	Creation()
	var command, product, address string
	var quantity int

	for {
		fmt.Scan(&command)
		switch command {
		case "ADD":
			fmt.Scan(&product, &quantity, &address)
			add(product, quantity, address)
		case "REMOVE":
			fmt.Scan(&product, &quantity, &address)
			remove(product, quantity, address)
		case "INFO":
			info()
		case "EXIT":
			return
		default:
			fmt.Println("Такой команды не существует")
		}
	}
}
